### DISCUSSION TITLE
Avoiding php-fpm restart with Apache

### DISCUSSION THREAD

**Discussion by Navi2016** [deployphp/deployer#3503] [Help needed] [answered]
Hi,

I read about the fixes to avoid the php-fpm reload/restart in this doc:
https://deployer.org/docs/7.x/avoid-php-fpm-reloading

But is there also a known fix for Apache 2.4?

Just using php-fpm restart after a deployment works fine, but is not an option for production.
Simply clearing the opcache does not seem to do enough!

**Comment by NeuralClone** [comment]
I realize I'm a bit late in responding to this, but I ran into this exact same problem recently and have a possible solution. Apache doesn't appear to have support for this out of the box. If it does, I haven't been able to find an officially supported solution.

However, the [Etsy realdoc](https://github.com/etsy/mod_realdoc) Apache module was written for this purpose. There isn't much documentation for it. But it's pretty easy to use and does the trick in my case.

**Reply by rajpt** [reply]
Same issue, for apache i need to reload php-fpm to pickup latest deployment from symlink. Any suggestions? Thank you.

**Reply by NeuralClone** [reply]
The only option I've found for Apache is the [Etsy realdoc](https://github.com/etsy/mod_realdoc) module I mentioned above. It provides functionality similar to the Nginx `$realpath_root` variable. You'll need root access to compile and install it.

If you have a cPanel managed server, configuration is a little more complicated. You'll need to reconfigure the document root for your primary domain (defaults to `/home/<user>/public_html`) or the configuration outlined in the `realdoc` documentation won't work.

**Reply by rajpt** [reply]
Thank you, i am looking something similar to Nginx in apache.

**Comment by khoibv** [comment]
After searching, I decided to reload `php-fpm` with Apache.
If you want to allow `deployer` to reload `php-fpm` service only (but cannot reload another services):

```
echo 'deployer ALL= NOPASSWD: /bin/systemctl reload php8.2-fpm' > /etc/sudoers.d/91-deployer-user
```

**Comment by stephanlucas** [comment]
Setting `opcache.revalidate_path = 1` in `php.ini` or `.user.ini` mostely works with apache. 

(Found this cheat a while ago here: [PHP opcache reset + symlink-style deployment - Stack Overflow](https://stackoverflow.com/questions/23737627/php-opcache-reset-symlink-style-deployment/52869752))

**Reply by NeuralClone** [reply]
Oh, interesting. I'll have to look into that some more. Thank you!

**Comment by bilogic** [comment]
Any new methods? This seems to do the job https://github.com/gordalina/cachetool

**Comment by NeuralClone** [comment, marked as answer]
Admittedly, I'm still trying to solve this problem over a year later since the other methods mentioned in this discussion (including my own suggestion) haven't worked very well.

The Etsy `realdoc` module *should* solve the problem, but configuring it correctly caused more problems than it was worth in a WHM/cPanel environment. It required making some risky changes to how cPanel auto-generates configuration files. There's a `--graceful` option when using the WHM/cPanel script to restart PHP-FPM, and that appears to not drop requests or cause them to fail.

@bilogic Conceptually, this tool looks like a strong contender to solve this problem. Thanks for the suggestion!

**Edit:** I just saw that there's already a Deployer recipe for this. I tried it out and it does exactly what I was looking for. Nice!

https://deployer.org/docs/7.x/contrib/cachetool

**Comment by sts-ryan-holton** [comment]
Struggling with the same thing, I have to keep rebooting the server which isn't great. What's the most current appropriate solution for this? I'm using Laravel and Laravel Octane on my Centos 9 Stream server

**Comment by NeuralClone** [comment]
@sts-ryan-holton The current recommendation in the docs is to set `opcache.revalidate_path=1` in `php.ini`:

https://github.com/deployphp/deployer/blob/master/docs/avoid-php-fpm-reloading.md

I wasn't able to get that to work reliably in a cPanel managed environment (AlmaLinux on a VPS in my case) and ended up using [CacheTool](https://github.com/gordalina/cachetool) as per [bilogic's suggestion](https://github.com/deployphp/deployer/discussions/3503#discussioncomment-10155168) above.

It can be easily integrated as a Deployer task:

https://deployer.org/docs/7.x/contrib/cachetool

I'm using Symfony 7.1 and Sulu 2.6. Either approach should work with Laravel as well.

**Comment by rajpt** [comment]
I endup doing this and it seems to be working fine in my app.

```
// restart services
task('restart:services', function () {
    // It doesn't automatically pick up the latest release symlink
    run('sudo systemctl reload php-fpm');
});
```

`before('deploy:success', 'restart:services');`
